### DISCUSSION TITLE
Use of UTF8 in program

### DISCUSSION THREAD

**Discussion by AlanoDel** [Oldes/Rebol3#81] [General]
Hi!
I have scripts in R2, but they not accept UTF8.
I need UTF8 for some languages, so my question is "is it possible to show and analyse text in UTF8?"
Thanks in advance,
Alain

**Comment by Oldes** [comment]
Could you provide some example? I am not sure, if I understand your question.

**Reply by AlanoDel** [reply]
I modified a program written by Allen Kamp.
The script load a file which contains words, in ASCII one per line, then choose randomly a word of length 7 and use the letters of the word to find all words length 3 to 7 which contains these letters. 
The script then display, by length, locations for the words which should be devined by the user.
As Rebol 2 was unable to use UTF8 coding I had to change accentd letters by a combination like cx for ĉ.

**Reply by AlanoDel** [reply]
How to send he script and the data files

**Reply by Oldes** [reply]
You can use gist: https://gist.github.com/AlanoDel

**Comment by Oldes** [comment]
If you need support for Rebol2, then I cannot help you. Rebol2 is out of my control. It was not Unicode aware. Cannot you use Rebol3 instead?

**Reply by AlanoDel** [reply]
Yes I want to use Rebols3 to have access to UTF8

**Reply by rgchris** [reply]
It was possible to work with UTF-8 in Rebol 2: strings and binary values were interchangeable 8-bit series and thus you had the freedom to put any old 8-bit sequence together—including UTF-8 (or Latin 1 or whatever). You needed separate tools to verify (one such was built-in: INVALID-UTF?), manipulate, and measure them (not difficult, each starting byte of a character sequence denotes a specific length) or to convert individual characters.

Rebol 3 introduced 16-bit strings (UTF-16). Of note is that 16-bit strings are limited to U+0 to U+FFFD where Unicode specifies characters ranging from U+0 to U+10FFFF (it was possible to store higher ranges in Rebol 3 Alpha strings as [surrogate pairs](https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF), though I'm not certain if this is standardised behaviour and is fraught with issues—and not permitted anyhow in this version of Rebol 3). However, it is still possible to work with UTF-8 sequences as binary values in Rebol 3 using some of the same techniques as Rebol 2 (INVALID-UTF? is still in there!), just don't expect them to be fully compatible with STRING!.

There are tradeoffs on basing a string datatype on UTF-16 vs. UTF-8 (predictable lengths vs. more widespread, more complete, arguably more efficient). It would seem that UTF-8 is the more popular, enduring, interoperable choice.

***Some data points:***

A single character above U+FFFD

    #"🙂"
    #1F642  ; Hex value
    128578  ; Numeric value
    #{F09F9982}  ; UTF-8 encoding
    #{D83DDE42}  ; UTF-16 (surrogate pair) encoding

A composite character above U+FFFD

    "🏴󠁧󠁢󠁳󠁣󠁴󠁿"
    #1F3F4 #E0067 #E0062 #E0073 #E0063 #E0074 #E007F  ; Hex values
    127988 917607 917602 917619 917603 917620 917631  ; Numeric values
    #{F09F8FB4F3A081A7F3A081A2F3A081B3F3A081A3F3A081B4F3A081BF}  ; UTF-8 encoding
    #{D83CDDF4D83CDF67D83CDF62D83CDF73D83CDF63D83CDF74D83CDF7F}  ; UTF-16 (surrogate pair) encoding

**Reply by Oldes** [reply]
@rgchris are you proposing something?

Here are related issues:
https://github.com/Oldes/Rebol-issues/issues/2024
https://github.com/Oldes/Rebol-issues/issues/2618

Problem is, that to add full Unicode support is VERY huge task (it does not matter if done with optional UCS-4 internal encoding or using UTF-8 everywhere).

And it should be noted, that I'm not an _emoticon lover_... For example in Red using a single emoticon in a string forces 4x higher memory use (as the string must be internally converted to use 32bits per char).

**Reply by Oldes** [reply]
@rgchris as you probably used Ren-C, where is already used _UTF-8 everywhere_, have you noticed any issues with it? Because so far I'm inclined to use it as well, instead of optional UCS-4. In many places it would simplify the code.

**Reply by rgchris** [reply]
@oldes Not proposing anything—writing as much to answer the OP question (I know it's old) as to restate my own understanding of how things *are* as opposed to how they might be: I haven't used R3 in a while and I'm trying to get reacquainted. I appreciate your emoticon skepticism 🙂, unfortunately they are pervasive across many domains and are broadly entwined with contemporary culture 😡.

I guess whether you decide to tackle UTF-8 depends on your priorities—as I mentioned, you can still work with UTF-8 fully within BINARY! values, so it's not a show-stopper. On the other hand, it would seem to be the future and might be a source of limitation if you don't address it. I have not used Ren-C with *UTF-8 everywhere* to any great extent (there were other semantic/language changes introduced concurrently that I couldn't really work with so went back to using R2).

**Reply by Oldes** [reply]
Getting closer...
![image](https://github.com/user-attachments/assets/2b4f9cfa-000e-4910-9f97-9652ac72f621)
...but still a long way to the finish.

**Reply by Oldes** [reply]
@rgchris do you have any opinion how to deal with uncomplete codepoints when converting binary to string? It looks that Red and Ren-C throws an error, but then one have almost no way how to convert at least part of the text :/

**Reply by Oldes** [reply]
On the other side, there is still `invalid-utf?` function, which can be used to find the location of the invalid codepoint and so read at least something. So maybe error is the way to go. Also the position could be in the error itself.

**Reply by rgchris** [reply]
Error is probably best, and yes—pinpointing the location of the error is useful. Some form of 'sanitize' method might also be useful for converting text with mixed encodings; though I think that might be better left to the user or user modules (I suspect it's rarer than it used to be).

There's some good testing resources here:

* [UTF-8 and Unicode FAQ for Unix/Linux](https://www.cl.cam.ac.uk/~mgk25/unicode.html#examples)
* [UTF-8 decoder capability and stress test](https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt)

**Comment by Oldes** [comment]
For text transcoding in Rebol3 there is `iconv` native:
```rebol
>> bin: to-binary "ábč"
== #{C3A162C48D} ;= text as utf8

>> iconv/to bin 'utf8 'cp1250
== #{E162E8}

>> iconv/to bin 'utf8 'cp1252
== #{E16263} ;= note that the č is converted to c as there is no č in this codepage!
```
In case you have a source in some known codepage and want to load it in Rebol3, you can do it like:
```
>> to string! iconv/to #{E162E8} 'cp1250 'utf8
== "ábč"
```
where instead of `#{E162E8}` would be `read %some.txt`.

Actually it can be just:
```
>> iconv #{E162E8} 'cp1250
== "ábč"
```
As the utf8 target is the default codepage.

Supported codepages can be found [in the source code](https://github.com/Oldes/Rebol3/blob/ea71e78ac989cbf99036d577f8db84b2ba0c8694/src/core/u-iconv.c#L59-L501).

**Comment by Oldes** [comment]
Unfortunately I don't have any script, which could guess, which encoding is used in an unknown text.

**Comment by Oldes** [comment]
There are also `ascii?` and `latin1?` natives which you may find useful...
```rebol

>> latin1? "abí" ;; Returns TRUE if value or string is in Latin-1 character range (below 256). 
== true

>> ascii? "abí" ;; Returns TRUE if value or string is in ASCII character range (below 128). 
== false
```
But as these requires string, the use may be limited, because conversion from binary to string expects UTF8 by default (if not used `iconv`).

**Comment by Oldes** [comment]
From version 3.19.5 Rebol is using UTF-8 encoded strings internally. https://github.com/Oldes/Rebol3/commit/a93ae36d1157b5b8e11dd8885245694c140e70af

This was significant change and so there may be some overlooked bugs.

**Reply by rgchris** [reply]
I don't know if this is a related bug or not, but this is a bit weird:

```
>> sort {"&^@}
== {^"&}

>> mold/all sort {"&^@}   
== {{^^"&}^@}
```

**Reply by Oldes** [reply]
Looks like a bug caused by the null char in the string.

**Reply by Oldes** [reply]
It's [fixed now](https://github.com/Oldes/Rebol3/commit/22ce4eab2694e8c59beab49828995a39572b3089). Btw.. it would be better to report bugs as issues. Not in discussions.
